src/inbox/rx.rs

```diff
@@ -0,0 +1,248 @@
+    fn new(rx: Receiver<A>) -> Self {
+        ReceiveFuture(ReceiveFutureInner::New(rx))
+    }
+}
+
+enum ReceiveFutureInner<A> {
+    New(Receiver<A>),
+    Waiting {
+        rx: Receiver<A>,
+        waiting: Arc<Spinlock<WaitingReceiver<A>>>,
+    },
+    Complete,
+}
+
+impl<A> Future for ReceiveFuture<A> {
+    type Output = ActorMessage<A>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<ActorMessage<A>> {
+        match mem::replace(&mut self.0, ReceiveFutureInner::Complete) {
+            ReceiveFutureInner::New(rx) => match rx.try_recv() {
+                Ok(message) => Poll::Ready(message),
+                Err(waiting) => {
+                    // Start waiting. The waiting receiver should be immediately polled, in case a
+                    // send operation happened between `try_recv` and here, in which case the
+                    // WaitingReceiver would be fulfilled, but not properly woken.
+                    self.0 = ReceiveFutureInner::Waiting { rx, waiting };
+                    self.poll_unpin(cx)
+                }
+            },
+            ReceiveFutureInner::Waiting { rx, waiting } => {
+                {
+                    let mut inner = waiting.lock();
+
+                    match inner.wake_reason.take() {
+                        // Message has been delivered
+                        Some(reason) => {
+                            return Poll::Ready(match reason {
+                                WakeReason::StolenMessage(msg) => msg.into(),
+                                WakeReason::BroadcastMessage => rx.pop_broadcast().unwrap(),
+                                WakeReason::Shutdown => ActorMessage::Shutdown,
+                                WakeReason::Cancelled => {
+                                    unreachable!("Waiting receive future cannot be interrupted")
+                                }
+                            })
+                        }
+                        // Message has not been delivered - continue waiting
+                        None => inner.waker = Some(cx.waker().clone()),
+                    }
+                }
+
+                self.0 = ReceiveFutureInner::Waiting { rx, waiting };
+                Poll::Pending
+            }
+            ReceiveFutureInner::Complete => Poll::Pending,
+        }
+    }
+}
+
+impl<A> ReceiveFuture<A> {
+    /// Cancel the receiving future, returning a message if it had been fulfilled with one, but had
+    /// not yet been polled after wakeup. Future calls to `Future::poll` will return `Poll::Pending`,
+    /// and `FusedFuture::is_terminated` will return `true`.
+    ///
+    /// This is important to do over `Drop`, as with `Drop` messages may be sent back into the
+    /// channel and could be observed as received out of order, if multiple receives are concurrent
+    /// on one thread.
+    #[must_use = "If dropped, messages could be lost"]
+    pub(crate) fn cancel(&mut self) -> Option<ActorMessage<A>> {
```

Comment:
Is this meant to be `pub(crate)`?

Reply:
For now, yes, since inbox::SendFuture won't be exposed directly - it will be wrapped in the other send future. That being said, this one should get a different name to differentiate it.